opensquirrel/passes/mapper/mip_mapper.py

```diff
@@ -0,0 +1,145 @@
+# OpenQL MIP-Like Mapper
+from __future__ import annotations
+
+from typing import Any, Optional
+
+import numpy as np
+from numpy.typing import NDArray
+from scipy.optimize import Bounds, LinearConstraint, milp
+
+from opensquirrel.ir import IR, Instruction, Qubit
+from opensquirrel.passes.mapper.general_mapper import Mapper
+from opensquirrel.passes.mapper.mapping import Mapping
+
+
+def compute_distance_matrix(connectivity: dict[str, list[int]], n: int) -> NDArray[np.int_]:
+    dist = np.full((n, n), np.inf)
+    np.fill_diagonal(dist, 0)
+    for i_str, neighbors in connectivity.items():
+        i = int(i_str)
+        for j in neighbors:
+            dist[i, j] = 1
+    for k in range(n):
+        for i in range(n):
+            for j in range(n):
+                if dist[i, j] > dist[i, k] + dist[k, j]:
+                    dist[i, j] = dist[i, k] + dist[k, j]
+    dist[dist == np.inf] = 999999
+    return dist.astype(int)
+
+
+class MIPMapper(Mapper):
+    def __init__(
+        self,
+        qubit_register_size: int,
+        connectivity: dict[str, list[int]],
+        timeout: Optional[float] = None,  # noqa: UP045
+        **kwargs: Any,
+    ) -> None:
+        self.qubit_register_size = qubit_register_size
+        self.connectivity = connectivity
+        self.timeout = timeout
+        self.distance_matrix = compute_distance_matrix(connectivity, qubit_register_size)
+        super().__init__(qubit_register_size, None, **kwargs)
+
+    def map(self, ir: IR) -> Mapping:  # noqa: C901
+        """
+        Find an initial mapping of virtual qubits to physical qubits that minimizes
+        the sum of distances between mapped operands of all two-qubit gates, using
+        Mixed Integer Programming (MIP).
+
+        This method formulates the mapping as a linear assignment problem, where the
+        objective is to minimize the total "distance cost" of executing all two-qubit
+        gates, given the hardware connectivity.
+
+        Args:
+            ir (IR): The intermediate representation of the quantum circuit to be mapped.
+
+        Returns:
+            Mapping: An object representing the mapping from virtual to physical qubits.
+
+        Raises:
+            RuntimeError: If the MIP solver fails to find a feasible mapping or times out.
+            RuntimeError: If the number of virtual qubits exceeds the number of physical qubits.
+        """
+        n = self.qubit_register_size
+        distance = self.distance_matrix
+
+        # Check if the number of virtual qubits exceeds the number of physical qubits
+        max_virtual_index = max(
+            (
+                q.index
+                for stmt in getattr(ir, "statements", [])
+                if isinstance(stmt, Instruction)
+                for q in getattr(stmt, "arguments", [])
+                if isinstance(q, Qubit)
```

Comment:
Use more descriptive variable names :)
```suggestion
        max_virtual_index = max(
            (
                q.index
                for statement in getattr(ir, "statements", [])
                if isinstance(statement, Instruction)
                for q in getattr(statement, "arguments", [])
                if isinstance(q, Qubit)
```